### PR TITLE
Instrument string comparisons in kotlin

### PR DESCRIPTION
**Pull request by kyakdan**

### CODE REVIEW

File: src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java [hunk 1]
```diff
@@ -213,6 +213,163 @@ private static void regionMatchesInternal(
...
+  @MethodHook(type = HookType.AFTER, targetClassName = "kotlin.text.StringsKt ",
+      targetMethod = "startsWith")
+  @MethodHook(type = HookType.AFTER, targetClassName = "kotlin.text.StringsKt ",
+      targetMethod = "startsWith$default")
+  public static void
+  startsWithKt(
+      MethodHandle method, Object alwaysNull, Object[] arguments, int hookId, Boolean returnValue) {
+    if (!returnValue) {
+      TraceDataFlowNativeCallbacks.traceStrstr(
+          (String) arguments[0], (String) arguments[1], hookId);
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java:272]: We should either check the types before casting or fix the method descriptors.

Comment (kyakdan) [reply src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java:272]: I'm still finalizing the PR and fixing this among other issues

Comment (kyakdan) [reply src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java]: @fmeum The PR can be reviewed now.

File: src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java [hunk 2]
```diff
@@ -213,6 +210,222 @@ private static void regionMatchesInternal(
...
+  @MethodHook(type = HookType.AFTER, targetClassName = "kotlin.text.StringsKt ",
+      targetMethod = "indexOfAny$default")
+  @MethodHook(type = HookType.AFTER, targetClassName = "kotlin.text.StringsKt ",
+      targetMethod = "lastIndexOfAny")
+  @MethodHook(type = HookType.AFTER, targetClassName = "kotlin.text.StringsKt ",
+      targetMethod = "lastIndexOfAny$default")
+  public static void
+  indexOfAnyKt(
+      MethodHandle method, Object alwaysNull, Object[] arguments, int hookId, Integer returnValue) {
+    if (arguments[0] instanceof CharSequence && returnValue == -1) {
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java:385]: We don't check the type of `arguments[1]` here.

File: src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java [hunk 3]
```diff
@@ -213,6 +210,222 @@ private static void regionMatchesInternal(
...
+  @MethodHook(type = HookType.AFTER, targetClassName = "kotlin.jvm.internal.Intrinsics ",
+      targetMethod = "areEqual")
+  @MethodHook(
+      type = HookType.AFTER, targetClassName = "kotlin.text.StringsKt ", targetMethod = "equals")
+  @MethodHook(type = HookType.AFTER, targetClassName = "kotlin.text.StringsKt ",
+      targetMethod = "equals$default")
+  public static void
+  equalsKt(MethodHandle method, Object alwaysNull, Object[] arguments, int hookId,
+      Boolean equalStrings) {
+    if (arguments[0] instanceof String && arguments[1] instanceof String && !equalStrings) {
```
Comment (fmeum) [inline src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java:224]: Not sure whether we have been doing this consistently with existing hooks, but we should probably also check the size of `arguments` in these hooks.

File: src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java (hunk 2 above)
Comment (kyakdan) [reply src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java:385]: There are two possible types for `arguments[1]`, and we check the type in `guideTowardContainmentOfFirstElement`.

File: src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java (hunk 3 above)
Comment (kyakdan) [reply src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java]: We haven't! I've gone through the CMP hooks where we don't specify a method descriptor and made sure that we check the size of `arguments` before accessing the array.

Comment (fmeum) [reply src/main/java/com/code_intelligence/jazzer/runtime/TraceCmpHooks.java]: Thanks! They all look safe now.

**Review by fmeum — APPROVED** [review]

**Pull request merged**